### PR TITLE
[analysis][NFC] Use C++20 concepts for Lattice

### PR DESCRIPTION
**Pull request by tlively**
Replace the static assertions ensuring that Lattice types have the necessary
operations with a C++20 concept called `Lattice`. To avoid name conflicts with
the new concept, rename existing type parameters named `Lattice` to `L`. When
not building with C++20, `Lattice` is a macro that resolves to `typename` so the
code continues compiling and has the same behavior, but without any eager checks
of the requirements on lattices.

Add a new C++20 builder to CI to ensure that future changes compile with both
C++17 and C++20. Once we switch to C++20 by default, the new builder can be
removed.

### CODE REVIEW

**Comment by tlively** [comment]
Current dependencies on/for this PR:
* main
  * **PR #6027** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6027" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/6027?utm_source=stack-comment).

File: src/analysis/lattice.h [hunk 1]
```diff
@@ -23,35 +23,33 @@ inline LatticeComparison reverseComparison(LatticeComparison comparison) {
...
-                      const typename Lattice::Element&>::value;
-template<typename Element>
-constexpr bool has_makeLeastUpperBound =
-  std::is_invocable_r<void,
-                      decltype(&Element::makeLeastUpperBound),
-                      Element,
-                      const Element&>::value;
-template<typename Element>
-constexpr bool has_isTop =
-  std::is_invocable_r<bool, decltype(&Element::isTop), const Element>::value;
```
Comment (kripken) [inline src/analysis/lattice.h:46]: What happened to `isTop` in C++20?

**Review by kripken — COMMENTED** [review]
First time I read Concepts code... :thinking:

File: src/analysis/lattice.h [hunk 2]
```diff
@@ -23,35 +23,33 @@ inline LatticeComparison reverseComparison(LatticeComparison comparison) {
...
+    lattice.compare(constElem, constElem)
+  } noexcept -> std::same_as<LatticeComparison>;
+  // We need to be able to get the least upper bound of two elements and know
+  // whether any change was made.
+  { elem.makeLeastUpperBound(constElem) } noexcept -> std::same_as<bool>;
+};
+
+#else
+
+#define Lattice typename
```
[the remark this answers is not in the document]
Comment (tlively) [reply src/analysis/lattice.h:50]: yup 🎉

File: src/analysis/lattice.h (hunk 1 above)
Comment (tlively) [reply src/analysis/lattice.h:46]: I don't think we need it for anything, so I didn't bother adding it back into the requirements.

File: src/analysis/lattice.h [hunk 3]
```diff
@@ -23,35 +23,33 @@ inline LatticeComparison reverseComparison(LatticeComparison comparison) {
...
-
-template<typename Lattice>
-constexpr bool is_lattice = has_getBottom<Lattice>&& has_compare<Lattice>&&
-  has_makeLeastUpperBound<typename Lattice::Element>&& has_isTop<
-    typename Lattice::Element>&& has_isBottom<typename Lattice::Element>;
+#if __cplusplus >= 202002L
+
+#include <concepts>
+
+template<typename T>
```
[the remark this answers is not in the document]
Comment (tlively) [reply src/analysis/lattice.h:30]: I can change it `L` for consistency.

**Review by kripken — APPROVED** [review]

**Review by ashleynh — APPROVED** [review]

**Pull request merged**